src/puppetlabs/puppetdb/config.clj

```diff
@@ -153,12 +154,23 @@
    :subprotocol "hsqldb"
    :subname     (format "file:%s;hsqldb.tx=mvcc;sql.syntax_pgs=true" (io/file vardir "db"))})
 
-(defn defaulted-db-connection
-  "Adds each of `hsql-default-connection` keypairs to `database` if classname, subprotocol and subname
-   are not found in the `database` config"
+(defn validate-db-settings
+  "Throws a {:type ::configuration-error :message m} exception
+  describing the required additions if the [database] configuration
+  doesn't specify classname, subprotocol and subname, all of which are
+  now required."
   [{:keys [global database] :as config :or {database {}}}]
-  (cond-> config (not-any? database [:classname :subprotocol :subname])
-    (update :database merge (hsql-default-connection (:vardir global)))))
+  (when (not-every? database [:classname :subprotocol :subname])
+    (let [default (hsql-default-connection (:vardir global))
+          setting-needed #(when-not (% config)
+                            (format "\n  %s = %s" (name %) (% default)))
+          msg (str "database configuration is now required. "
+                   "Please add this to the [database] section of your config file:"
```

Comment:
We should qualify this with something like 'if you want to use hsqldb, which is deprecated, and you shouldn't use'.